docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
 
 is diagnosed as a label in which the user forgot the **label** keyword.
 
+## Update with Failure Statement
+
+An update statement using `:-`   instead of `:=` exits execution of a method immediately if a failure is reported. This is a form of exceptional return from the calling method.
+
+To use this form of update,
+   * the caller must have a first output value whose type is failure-compatible
+   * if the RHS of the update-with-failure statement is a method call, the first output value of the callee must be failure-compatible
+   * if instead the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
+
+A failure-compatible type is a type that has the following:
+   * a predicate `predicate IsFailure()`
+   * a method `method PropagateFailure() returns _X_ `: The type _X_ must be a failure-compatible type and must be the type of the first output of the caller.
+   * a method `method Extract() returns _T_` where _T_ is some type for the value carried by the type.
+
+TODO: Why should PropagateFailure and Extract be functions?
+
+### RHS with method call
+
+A common example of a failure-compatible type is the following:
+```
+datatype Outcome<T> =
+| Success(value: T)
+| Failure(error: string)
+{
+  predicate method IsFailure() {
+    this.Failure?
+  }
+  function method PropagateFailure(): Outcome<U>
+    requires IsFailure()
+  {
+    Failure(this.error); // this is Outcome<U>.Failure(...)
+  }
+  function method Extract(): T
+    requires !IsFailure()
+  {
+    this.value
+  }
+}
+```
+
+The use of this statement is best explained by an example:
+```
+method callee(i: int) returns (r: Outcome<nat>, v: int)
```

Comment:
Rename `caller` to `Caller` and `callee` to `Callee`.

Reply:
Here and elsewhere